adview-manager/src/lib.rs

```diff
@@ -0,0 +1,338 @@
+    pub targeting_score: TargetingScore,
+}
+
+impl Unit {
+    pub fn new(by_price: UnitByPrice, targeting_score: TargetingScore) -> Self {
+        Self {
+            unit: by_price.unit,
+            channel_id: by_price.channel_id,
+            validators: by_price.validators,
+            min_targeting_score: by_price.min_targeting_score,
+            min_per_impression: by_price.min_per_impression,
+            targeting_score,
+        }
+    }
+}
+
+pub fn apply_selection(campaigns: &[MarketChannel], options: &AdViewManagerOptions) -> Vec<Unit> {
+    let eligible = campaigns.iter().filter(|campaign| {
+        options
+            .accepted_states
+            .contains(&campaign.status.status_type)
+            && campaign
+                .spec
+                .active_from
+                .map(|datetime| datetime < Utc::now())
+                .unwrap_or(true)
+            && campaign.deposit_asset == options.whitelisted_token
+            && campaign.spec.min_per_impression >= options.min_per_impression
+    });
+
+    let mut units: Vec<UnitByPrice> = eligible
+        .flat_map(|campaign| {
+            let mut units = vec![];
+            for ad_unit in campaign.spec.ad_units.iter() {
+                let unit = UnitByPrice {
+                    unit: ad_unit.clone(),
+                    channel_id: campaign.id.clone(),
+                    validators: campaign.spec.validators.clone(),
+                    min_targeting_score: ad_unit
+                        .min_targeting_score
+                        .or(campaign.spec.min_targeting_score)
+                        .unwrap_or(0.into()),
+                    min_per_impression: campaign.spec.min_per_impression.clone(),
+                };
+
+                units.push(unit);
+            }
+
+            units
+        })
+        .collect();
+
+    // Sort
+    units.sort_by(|b, a| a.min_per_impression.cmp(&b.min_per_impression));
+    units.truncate(options.top_by_price);
+
+    let units = units.into_iter().filter(|unit| {
+        options
+            .whitelisted_type
+            .as_ref()
+            .map(|whitelisted_type| {
+                whitelisted_type != &unit.unit.ad_type
+                    && !(options.disabled_video && is_video(&unit.unit))
+            })
+            .unwrap_or(false)
+    });
+
+    let mut by_score: Vec<Unit> = units
+        .collect::<Vec<UnitByPrice>>()
+        .into_iter()
+        .filter_map(|by_price| {
+            let targeting_score =
+                calculate_target_score(&by_price.unit.targeting, &options.targeting);
+            if targeting_score >= options.min_targeting_score
+                && targeting_score >= by_price.min_targeting_score
+            {
+                Some(Unit::new(by_price, targeting_score))
+            } else {
+                None
+            }
+        })
+        .collect();
+    by_score.sort_by(|a, b| {
+        a.targeting_score
+            .partial_cmp(&b.targeting_score)
```

Comment:
why is this returning a Result? just curious

Reply:
Targeting score is f64, we can't directly compare them, but we can do a `PartialCmp`:
[Docs](https://doc.rust-lang.org/std/cmp/trait.PartialOrd.html#tymethod.partial_cmp)
```
let result = std::f64::NAN.partial_cmp(&1.0);
assert_eq!(result, None);
```